crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -6924,6 +6924,11 @@ impl<'db> TypeInferenceBuilder<'db> {
                         return Type::unknown();
                     };
 
+                    let bound_signature = signature.bind_self();
+                    if let Type::BoundMethod(_) = argument_type {
+                        signature = &bound_signature;
+                    }
+
                     Type::Callable(CallableType::new(db, signature.clone()))
```

Comment:
this should be slightly more performant, and looks a little cleaner :-)

```suggestion
                    let bound_signature = if argument_type.is_bound_method() {
                        signature.bind_self()
                    else {
                        signature.clone()
                    };

                    Type::Callable(CallableType::new(db, signature))
```

It requires adding a `Type::is_bound_method()` method, similar to this one:

https://github.com/astral-sh/ruff/blob/4571c5f56f660d04c89f2ac00bfb8c0159597561/crates/red_knot_python_semantic/src/types.rs#L578-L580